### ISSUE TITLE
Validation Error for Document.type and Parameter.name

### ISSUE THREAD

**Issue by MRichards99** [ral-facilities/datagateway-api#317]
**Description**:
The following requests give errors:

Request:
`http://{{datagateway-api}}/search-api/datasets/0-8401-1070-7?filter={"include": [{"relation": "documents"}]}`

Error from logs:
```bash
File "pydantic/main.py", line 406, in pydantic.main.BaseModel.__init__
pydantic.error_wrappers.ValidationError: 1 validation error for Document
type
  field required (type=value_error.missing)
```

Request:
`http://{{datagateway-api}}/search-api/documents/0-449-78690-0?filter={"include": [{"relation": "parameters"}]}`

Error from logs:
```bash
File "pydantic/main.py", line 406, in pydantic.main.BaseModel.__init__
pydantic.error_wrappers.ValidationError: 1 validation error for Parameter
name
  field required (type=value_error.missing)
```

My mappings file is the same the example on `feature/search-endpoints-#268` but I'll paste in the relevant mapping objects just in case:

```json
"Dataset": {
        "base_icat_entity": "Dataset",
        "pid": "doi",
        "title": "name",
        "isPublic": "createTime",
        "creationDate": "createTime",
        "size": "",
        "documents": {"Document": "investigation"},
        "techniques": {"Technique": "datasetTechniques.technique"},
        "instrument": {"Instrument": "datasetInstruments.instrument"},
        "files": {"File": "datafiles"},
        "parameters": {"Parameter": "parameters"},
        "samples": {"Sample": "sample"}
    },
    "Document": {
        "base_icat_entity": "Investigation",
        "pid": "doi",
        "isPublic": "releaseDate",
        "type": "type.name",
        "title": "name",
        "summary": "summary",
        "doi": "doi",
        "startDate": "startDate",
        "endDate": "endDate",
        "releaseDate": "releaseDate",
        "license": "",
        "keywords": "keywords.name",
        "datasets": {"Dataset": "datasets"},
        "members": {"Member": "investigationUsers"},
        "parameters": {"Parameter": "parameters"}
    },
  "Parameter": {
          "base_icat_entity": ["InvestigationParameter", "DatasetParameter"],
          "id": "id",
          "name": "type.name",
          "value": ["numericValue", "stringValue", "dateTimeValue"],
          "unit": "type.units",
          "dataset": {"Dataset": "dataset"},
          "document": {"Document": "investigation"}
      },
```

I think the mappings are right? I've gone through them using the ICAT schema and I can't see any issues with the field names. In both cases, the fields in ICAT are mandatory so I don't think its the same issue as #314. I've also used DataGateway API to query some of the data to ensure those fields are populated, which they are.

**Acceptance criteria**:
- [ ] Including the Document entity shouldn't cause an error, it should return data properly
- [ ] Including the Parameter entity shouldn't cause an error, it should return data properly
- [ ] Unskip relevant tests in `test_get_entity_by_pid.py` that have TODOs against them

**Comment by VKTB** [comment]
@MRichards99 Both, Document's `type` and Parameter's `name` field get their values from related entities in ICAT. If querying `/documents` returns data fine then these errors are probably thrown because the ICAT relations for the non-related fields of the related PaNOSC entities (Document and Parameter in this case) are not retrieved. Are you sure you have integrated the changes from my https://github.com/ral-facilities/datagateway-api/pull/304 PR?

When I was testing the endpoints the other day and integrated the changes from the above PR and `master`, I remember testing this case and not getting any errors.

**Comment by MRichards99** [comment]
I think so, I merged in master after you merged #304 (see https://github.com/ral-facilities/datagateway-api/pull/302/commits/efdc2a61fce57b44920f513dbb7f2c9d7dc8d3a5). If you look at the helper function for get by pid endpoint (https://github.com/ral-facilities/datagateway-api/blob/feature/search-endpoints-%23268/datagateway_api/src/search_api/helpers.py#L67), lines 90-97 contain calls to that code and I think they're correct?

**Comment by VKTB** [comment]
Ah, so, it looks like only the `get_icat_relations_for_panosc_non_related_fields` method, which gets the ICAT relations for the **non-related fields of the PaNOSC entity under query**, is being called. Remember that I moved the `FilterOrderHandler` class to `common` so that I can define the below method which deals with getting the ICAT relations for the **non-related fields of the PaNOSC related entities** (Document and Parameter in the case above)?
https://github.com/ral-facilities/datagateway-api/blob/fcc2e5729ec659c6ba72fe525fad541bd1948ac4/datagateway_api/src/common/filter_order_handler.py#L50-L52
Because of this, that method needs to be called too - you can do that between lines 103 and 104 like this:
`filter_handler.add_icat_relations_for_non_related_fields_of_panosc_related_entities(entity_name)`.

I remember saying that lines 90-97 could not be moved to the `FilterOrderHandler` class because it does not have access to the PaNOSC entity name, however looking at it now, the `add_icat_relations_for_non_related_fields_of_panosc_related_entities` too needed the PaNOSC entity name so I dealt with this by defining the method to take this as an argument. In addition to this, I see that a `FilterOrderHandler` object is always created even if the `filters` list is empty. As a result of all this, I think (needs testing) that we can move the logic in lines 90-97 in a new method (`add_icat_relations_for_panosc_non_related_fields`) across to the `FilterOrderHandler` class and then call this method before we call `add_icat_relations_for_non_related_fields_of_panosc_related_entities`. I do not know why this did not occur to me when I was working on that PR.

I notice that these same changes will need to be made to the `get_search` method in order for it to work as expected, however it seems like there is a bit of code duplication going on between that and the `get_with_pid` method. Conscious that you are still working on your PR and may have already a plan to do some refactoring when it comes to do this, but I thought I would suggest the idea of removing some of the duplicated code in `get_with_pid` and modifying its logic so that it calls the `get_search`. It's only a suggestion so I am not sure if it will work. I will add my suggested code to your PR shortly.

**Comment by MRichards99** [comment]
Closed now #302 has been merged
